File: modules/aidj/aidj相关/aidjmix-logic-rotation-and-logs/rotation/py/preset_scheduler.py
# rotation/py/preset_scheduler.py
# Logic-only preset rotation for Python workflows (no ports, no servers).

from dataclasses import dataclass
from typing import List, Optional, Dict, Tuple
import datetime as _dt

PresetName = str  # 'deep_minimal' | 'classic' | 'peak_warehouse' | 'hard_techno' | 'hypnotic'

@dataclass
class RotationSlot:
    start_hour: int  # 0..23
    end_hour: int    # 1..24
    preset: PresetName
    simple_head_tail: bool = False
    notes: str = ""

DEFAULT_ROTATION: List[RotationSlot] = [
    RotationSlot(0, 6,  'deep_minimal',  True,  'late night low-key flow'),
    RotationSlot(6, 12, 'classic',       False, 'daytime baseline'),
    RotationSlot(12,18, 'peak_warehouse',False, 'afternoon energy'),
    RotationSlot(18,22, 'peak_warehouse',False, 'prime time push'),
    RotationSlot(22,24, 'hypnotic',      True,  'late evening long layers'),
]

def _in_range(h: int, slot: RotationSlot) -> bool:
    if slot.start_hour < slot.end_hour:
        return slot.start_hour <= h < slot.end_hour
    return h >= slot.start_hour or h < slot.end_hour

def select_preset(local_dt: _dt.datetime, rotation: List[RotationSlot]=DEFAULT_ROTATION,
                  telemetry: Optional[Dict]=None) -> Dict:
    h = local_dt.hour
    slot = next((s for s in rotation if _in_range(h, s)), rotation[0])
    reasons = [f"time:{h}:00 in [{slot.start_hour},{slot.end_hour}) → {slot.preset}"]
    simple = bool(slot.simple_head_tail)

    if telemetry:
        recent_errors = telemetry.get('recentErrors', 0)
        dropout = telemetry.get('dropoutRate', 0.0)
        avg_bpm = telemetry.get('avgBpm', None)
        if recent_errors > 0:
            simple = True; reasons.append('recentErrors>0 → force simple_head_tail')
        if isinstance(dropout, (int,float)) and dropout > 0.05:
            simple = True; reasons.append('dropoutRate>5% → simplify transitions')
        if isinstance(avg_bpm, (int,float)):
            if avg_bpm >= 132 and slot.preset in ('deep_minimal','classic','hypnotic'):
                reasons.append(f'avgBpm={avg_bpm} → bias up')
                return {'preset':'peak_warehouse','simple_head_tail':simple,'reason':reasons}
            if avg_bpm <= 126 and slot.preset in ('peak_warehouse','hard_techno'):
                reasons.append(f'avgBpm={avg_bpm} → bias down')
                return {'preset':'classic','simple_head_tail':simple,'reason':reasons}

    return {'preset':slot.preset, 'simple_head_tail':simple, 'reason':reasons}

def minutes_until_next_change(local_dt: _dt.datetime, rotation: List[RotationSlot]=DEFAULT_ROTATION) -> int:
    h = local_dt.hour; m = local_dt.minute
    slot = next((s for s in rotation if _in_range(h, s)), rotation[0])
    end_h = slot.end_hour % 24
    minutes = ((end_h - h) % 24)*60 - m
    return minutes if minutes>0 else 60

# Usage (no ports):
# sel = select_preset(_dt.datetime.now(), DEFAULT_ROTATION, {'avgBpm':130, 'dropoutRate':0.02})
# args = ['--preset', sel['preset']] + (['--simple_head_tail'] if sel['simple_head_tail'] else [])

File: modules/aidj/aidj相关/aidjmix-logic-rotation-and-logs/rotation/ts/presetScheduler.ts
// rotation/ts/presetScheduler.ts
// Logic-only preset rotation (no ports, no servers).
// Choose a Techno preset by local time + (optional) telemetry, return flags for simpleHeadTail.
// Safe to import from any frontend/worker code without altering your Managers.

export type PresetName = 'deep_minimal' | 'classic' | 'peak_warehouse' | 'hard_techno' | 'hypnotic';

export interface RotationSlot {
  startHour: number;   // inclusive, 0..23 (local time)
  endHour: number;     // exclusive, 1..24 (local time)
  preset: PresetName;
  simpleHeadTail?: boolean; // true: only head-tail connections (no extra automation hints)
  notes?: string;
}

export interface Telemetry {
  // Optional realtime signals to bias selection (all optional)
  listeners?: number;        // concurrent listeners
  dropoutRate?: number;      // 0..1 (higher => simplify transitions)
  recentErrors?: number;     // number of failures in last window
  avgBpm?: number;           // current average BPM of nowplaying
}

export interface Selection {
  preset: PresetName;
  simpleHeadTail: boolean;
  reason: string[];
}

export const DEFAULT_ROTATION: RotationSlot[] = [
  { startHour: 0,  endHour: 6,  preset: 'deep_minimal',  simpleHeadTail: true,  notes: 'late night low-key flow' },
  { startHour: 6,  endHour: 12, preset: 'classic',       simpleHeadTail: false, notes: 'daytime baseline' },
  { startHour: 12, endHour: 18, preset: 'peak_warehouse',simpleHeadTail: false, notes: 'afternoon energy' },
  { startHour: 18, endHour: 22, preset: 'peak_warehouse',simpleHeadTail: false, notes: 'prime time push' },
  { startHour: 22, endHour: 24, preset: 'hypnotic',      simpleHeadTail: true,  notes: 'late evening long layers' },
];

function inRange(h: number, s: RotationSlot): boolean {
  if (s.startHour < s.endHour) return h >= s.startHour && h < s.endHour;
  // overnight wrap (not used in default but kept generic)
  return h >= s.startHour || h < s.endHour;
}

export function selectPreset(localDate: Date, rotation: RotationSlot[] = DEFAULT_ROTATION, telemetry?: Telemetry): Selection {
  const h = localDate.getHours();
  const slot = rotation.find(s => inRange(h, s)) ?? rotation[0];
  const reason: string[] = [`time:${h}:00 in [${slot.startHour},${slot.endHour}) → ${slot.preset}`];
  let simple = !!slot.simpleHeadTail;

  // Stability guardrails
  if (telemetry?.recentErrors && telemetry.recentErrors > 0) {
    simple = true; reason.push('recentErrors>0 → force simpleHeadTail');
  }
  if (typeof telemetry?.dropoutRate === 'number' && telemetry.dropoutRate > 0.05) {
    simple = true; reason.push('dropoutRate>5% → simplify transitions');
  }
  if (typeof telemetry?.avgBpm === 'number') {
    // If avg BPM > 132 and preset is mellow, bias to peak_warehouse
    if (telemetry.avgBpm >= 132 && (slot.preset === 'deep_minimal' || slot.preset === 'classic' || slot.preset === 'hypnotic')) {
      reason.push(`avgBpm=${telemetry.avgBpm} → bias up`);
      return { preset: 'peak_warehouse', simpleHeadTail: simple, reason };
    }
    // If avg BPM < 126 and preset is aggressive, relax
    if (telemetry.avgBpm <= 126 && (slot.preset === 'peak_warehouse' || slot.preset === 'hard_techno')) {
      reason.push(`avgBpm=${telemetry.avgBpm} → bias down`);
      return { preset: 'classic', simpleHeadTail: simple, reason };
    }
  }
  return { preset: slot.preset, simpleHeadTail: simple, reason };
}

// Optional helper: compute next change time (minutes until next slot)
export function minutesUntilNextChange(localDate: Date, rotation: RotationSlot[] = DEFAULT_ROTATION): number {
  const h = localDate.getHours();
  const m = localDate.getMinutes();
  const cur = rotation.find(s => inRange(h, s)) ?? rotation[0];
  const endH = cur.endHour % 24;
  const minutesToEnd = ((endH - h + 24) % 24) * 60 - m;
  return minutesToEnd <= 0 ? 60 : minutesToEnd;
}

// Usage (no port):
// const sel = selectPreset(new Date(), DEFAULT_ROTATION, { listeners, dropoutRate, recentErrors, avgBpm });
// client.requestAutoPlaylist(tracks, 60, 24, /* body can include */ sel);
